flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DefaultJobManagerRunnerRegistry.java

```diff
@@ -97,6 +106,14 @@ public JobManagerRunner unregister(JobID jobId) {
         return this.jobManagerRunners.remove(jobId);
     }
 
+    public void setMainThreadExecutor(ComponentMainThreadExecutor executor) {
+        mainThreadExecutor = executor;
+    }
+
+    public ComponentMainThreadExecutor getMainThreadExecutor() {
+        return mainThreadExecutor;
+    }
+
```

Comment:
hm, what about a 4th option where we implement the `DispatcherResourceCleanerFactory` using the Builder pattern? That would allow us to set the `JobManagerRunnerRegistry` after instantiating the `ResourceCleanerFactory` instance. 🤔

Reply:
You mean combing Factory and Builder patterns for `DispatcherResourceCleanerFactory`? How about passing the `JobManagerRunnerRegistry` along with `ComponentMainThreadExecutor` in the creator methods there?

```java
public ResourceCleaner createLocalResourceCleaner(
        ComponentMainThreadExecutor mainThreadExecutor,
+        JobManagerRunnerRegistry jobManagerRunnerRegistry);
```

```java
public ResourceCleaner createGlobalResourceCleaner(
        ComponentMainThreadExecutor mainThreadExecutor,
+        JobManagerRunnerRegistry jobManagerRunnerRegistry);
```

Reply:
yeah, I just noticed when thinking about it that using the Builder pattern doesn't make a difference. Essentially, it's about adding the `JobManagerRunnerRegistry` to the `create*ResourceCleaner` method signatures of the `ResourceCleanerFactory` which is a bit unfortunate.

But what we can do is making the main thread passing an implementation detail of the `DispatcherResourceCleanerFactory`: We could introduce a twin interface `LocallyCleanableWithMainThreadResource`:
```java
@FunctionalInterface
public interface LocallyCleanableInMainThreadResource {

    CompletableFuture<Void> localCleanupAsync(
            JobID jobId, Executor cleanupExecutor, Executor mainThreadExecutor);
}
```

That interface can be converted to a `LocallyCleanableResource` in `DispatcherResourceCleanerFactory`. We wouldn't have to change the `Dispatcher` instantiation in that case. WDYT?

Reply:
> But what we can do is making the main thread passing an implementation detail of the `DispatcherResourceCleanerFactory`: We could introduce a twin interface `LocallyCleanableWithMainThreadResource`:
> 
> ```java
> @FunctionalInterface
> public interface LocallyCleanableInMainThreadResource {
> 
>     CompletableFuture<Void> localCleanupAsync(
>             JobID jobId, Executor cleanupExecutor, Executor mainThreadExecutor);
> }
> ```
> 
> That interface can be converted to a `LocallyCleanableResource` in `DispatcherResourceCleanerFactory`. We wouldn't have to change the `Dispatcher` instantiation in that case. WDYT?

I didn't get this. Which classes would implement `LocallyCleanableInMainThreadResource`? Can you elaborate more?

Reply:
JobManagerRunnerRegistry would implement `LocallyCleanableInMainThreadResource` instead of `LocallyCleanableResource`. The conversion of the `LocallyCleanableInMainThreadResource#localCleanupAsync` callback to the `LocallyCleanableResource#localCleanupAsync` callback would then happen in `DispatcherResourceCleanerFactory#create[Loc|Glob]alResourceCleaner`.

Reply:
Done, LMK if you still want to collapse `DefaultJobManagerRunnerRegistry` and `OnMainThreadJobManagerRunnerRegistry`.